wazo_dird/database/queries/phonebook.py

```diff
@@ -303,35 +335,61 @@ def _get_contact(
 
     def _get_phonebook(
         self,
-        s: Session,
+        s: BaseSession,
         visible_tenants: builtins.list[str] | None,
         phonebook_key: PhonebookKey,
     ) -> Phonebook:
-        phonebook_filter = phonebook_key_to_predicate(phonebook_key)
-        if visible_tenants is None:
-            return phonebook_filter
-        elif not visible_tenants:
-            return text('false')
-        else:
-            filter_ = and_(phonebook_filter, Phonebook.tenant_uuid.in_(visible_tenants))
-            phonebook = s.query(Phonebook).filter(filter_).first()
-            if not phonebook:
-                raise NoSuchPhonebook(cast(dict, phonebook_key))
-            return phonebook
+        filter_ = phonebook_selector(visible_tenants, phonebook_key)
+
+        phonebook = s.query(Phonebook).filter(filter_).first()
+        if not phonebook:
+            raise NoSuchPhonebook(
+                cast(dict, phonebook_key), tenants_in_scope=visible_tenants
+            )
+        return phonebook
 
-    def _list_contacts(
+    def _query_contacts(
         self,
-        s: Session,
-        query: ColumnElement,
+        s: BaseSession,
         phonebook_key: PhonebookKey,
-        search: str | None,
-    ):
+        search: str | None = None,
+    ) -> Query:
         phonebook_filter = contact_phonebook_key_to_predicate(phonebook_key)
+        search_filter = contact_search_filter(
+            phonebook_key=phonebook_key, search=search
+        )
+        filter_ = and_(phonebook_filter, search_filter)
+        query = s.query(Contact).join(Phonebook).filter(filter_)
+        return query
+
+    def _query_contact_fields(
+        self,
+        s: BaseSession,
+        phonebook_key: PhonebookKey,
+        search: str | None = None,
+    ) -> Query:
+        phonebook_filter = contact_phonebook_key_to_predicate(phonebook_key)
+        search_filter = contact_search_filter(
+            phonebook_key=phonebook_key, search=search
+        )
```

Comment:
Compared to the old version, this introduces an unnecessary subquery. Can we avoid that?

Reply:
Actually, there's a logical difference, though I admit I'm not sure how the ORM is dealing with it. The subquery selects the contacts that apply based on their fields, which the main query pulls in whole. So a contact is represented by multiple rows in the main query, and we want to filter _per group_, not individual rows(we want to exclude the contacts that have no matching fields, not the fields that do not match). I'm not sure it's possible to do that without a subquery?

Reply:
I personally do not have a problem with subqueries if it makes the query easier to understand and modify. Unless there's a known problem I would not rework this query at the moment.